### PR TITLE
[GEOT-6512] URLCheckers normalize location

### PR DESCRIPTION
**Pull request by jodygarnett**
[GEOT-6512](https://osgeo-org.atlassian.net/browse/GEOT-6512) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

This is a follow up to https://github.com/geoserver/geoserver/pull/6916 which highlighted the need to account for normalization of locations prior to checking.

URLChecker.normalize(location) is introduced to:
- remove any redundant current directory or parent directory references.
- Converts ``file:/path`` locations to ``file:///path`` representation.

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

**Comment by jodygarnett** [comment]
Okay I am not enjoying this one at all ... I really wish to use the Java built-in URI and Path normalization rather than roll my own as I do not wish any possibility of error/inconsistency in a facility setup to address security vulnerabilities.

Glitches:

1. File URI / URL handling has two supported representations for ``file:/path`` and ``file:///path`` both of which are equally valid. The ``uri.normalize().toString()`` workflow produces inconsistant output depending on how the URI was constructed. When creating a URI using parameters or normalize() this will create ``file:/path`` (no host location), although when URI is created using a String it is willing to preserve the string resulting in ``file:///path`` (empty host location).

2. Windows ``\\`` vs Linux ``/`` will cause path.normalize().toString() to change ``/path/file`` into ``\\path\\file`` which may or may not be what people expect to check against?

File: modules/library/main/src/test/java/org/geotools/data/ows/URLCheckersTest.java [hunk 1]
```diff
@@ -72,18 +74,99 @@ public void testMultipleCheckers() {
...
+                "https://localhost:8080/geoserver/ne/ows?SERVICE=WFS&REQUEST=GetCapabilities&VERSION=1.1.0");
+
+        // valid normalization
+        URLCheckers.confirm("file:///data/../data/dem.tif");
+        URLCheckers.confirm("file:/data/../data/dem.tif");
+        URLCheckers.confirm("file:///data/./dem.tif");
+        URLCheckers.confirm("file:/data/./dem.tif");
+
+        // invalid normalization
+        assertThrows(URLCheckerException.class, () -> URLCheckers.confirm("file:///tmp"));
```
Comment (aaime) [inline modules/library/main/src/test/java/org/geotools/data/ows/URLCheckersTest.java:127]: Maybe also add one to confirm path traversal has been neutralized, e.g. ``file:///data/../tmp``

File: modules/library/main/src/test/java/org/geotools/data/ows/URLCheckersTest.java [hunk 2]
```diff
@@ -72,18 +74,99 @@ public void testMultipleCheckers() {
...
+    }
+
+    @Test
+    public void testNormalization() {
+        Assert.assertEquals(
```
Comment (aaime) [inline modules/library/main/src/test/java/org/geotools/data/ows/URLCheckersTest.java:132]: static import for assertEquals?

**Review by aaime — CHANGES_REQUESTED** [review]
Looks good to me. I see the normalization is a bit painful, but I would not be able to suggest a better approach.

**Comment by jodygarnett** [comment]
This PR must be merged at the same time as https://github.com/geoserver/geoserver/pull/6916 for the "Downstream integration build (GeoWebCache and GeoServer) / build (pull_request) " test to pass.

**Comment by jodygarnett** [comment]
Thanks @aaime feedback addressed, let me push up the changes to https://github.com/geoserver/geoserver/pull/6916 and we can stage all of this stuff for backport.

File: modules/library/main/src/test/java/org/geotools/data/ows/URLCheckersTest.java (hunk 1 above)
Comment (jodygarnett) [reply modules/library/main/src/test/java/org/geotools/data/ows/URLCheckersTest.java:127]: Added

**Review by ianturton — APPROVED** [review]

**Pull request merged**